lib/fhir_client/client.rb

```diff
@@ -440,309 +442,42 @@ def scrubbed_response_headers(result)
-        begin
-          response = @client.get(url, headers: headers)
-        rescue => e
-          if !e.respond_to?(:response) || e.response.nil?
-            # Re-raise the client error if there's no response. Otherwise, logging
-            # and other things break below!
-            FHIR.logger.error "GET - Request: #{url} failed! No response from server: #{e}"
-            raise # Re-raise the same error we caught.
-          end
-          response = e.response if e.response
-        end
-        req = {
-          method: :get,
-          url: url,
-          path: url.gsub(@base_service_url, ''),
-          headers: headers,
-          payload: nil
-        }
-        res = {
-          code: response.status.to_s,
-          headers: response.headers,
-          body: response.body
-        }
-        if url.end_with?('/metadata')
-          FHIR.logger.debug "GET - Request: #{req}, Response: [too large]"
-        else
-          FHIR.logger.debug "GET - Request: #{req}, Response: #{response.body.force_encoding('UTF-8')}"
-        end
-        @reply = FHIR::ClientReply.new(req, res, self)
-      else
-        headers.merge!(@security_headers) if @use_basic_auth
-        begin
-          response = @client.get(url, headers)
-        rescue RestClient::SSLCertificateNotVerified => sslerr
-          FHIR.logger.error "SSL Error: #{url}"
-          req = {
-            method: :get,
-            url: url,
-            path: url.gsub(@base_service_url, ''),
-            headers: headers,
-            payload: nil
-          }
-          res = {
-            code: nil,
-            headers: nil,
-            body: sslerr.message
-          }
-          @reply = FHIR::ClientReply.new(req, res, self)
-          return @reply
-        rescue => e
-          if !e.respond_to?(:response) || e.response.nil?
-            # Re-raise the client error if there's no response. Otherwise, logging
-            # and other things break below!
-            FHIR.logger.error "GET - Request: #{url} failed! No response from server: #{e}"
-            raise # Re-raise the same error we caught.
-          end
-          response = e.response
-        end
-        if url.end_with?('/metadata')
-          FHIR.logger.debug "GET - Request: #{response.request.to_json}, Response: [too large]"
-        else
-          FHIR.logger.debug "GET - Request: #{response.request.to_json}, Response: #{response.body.force_encoding('UTF-8')}"
-        end
-        response.request.args[:path] = response.request.args[:url].gsub(@base_service_url, '')
-        headers = response.headers.each_with_object({}) { |(k, v), h| h[k.to_s.tr('_', '-')] = v.to_s; h }
-        res = {
-          code: response.code,
-          headers: scrubbed_response_headers(headers),
-          body: response.body
-        }
-
-        @reply = FHIR::ClientReply.new(response.request.args, res, self)
-      end
+      @reply = perform_request(method: :get, path: path, headers: headers)
```

Comment:
Each of the HTTP verbs has its own method in the old implementation. They contain a lot of duplicated and overly-complicated code. They don't all do exactly the same thing, so my refactoring here _does_ change behavior. That could be a problem, but it seems unlikely to me. The differences in the different methods seem arbitrary to me (but, of course, I could be wrong).

Reply:
I trust you